core/src/test/java/org/neo4j/ogm/persistence/examples/pizza/PizzaIntegrationTest.java

```diff
@@ -390,17 +390,65 @@ public void shouldBeAbleToModifyPropertiesAndRelsWithinSingleSave() {
+
+        // What's happening?
+        // when the pizza is saved, a 2-statement cypher request is made. The first creates the
+        // new Crust object, the second one updates the Pizza object, setting the crust relationship
+        // and updating the property.
+
+        // However as mentioned, the object in the session at the save point, which represents the pizza, is not
+        // in fact the pizza, but the earlier, loadedPizza - which hasn't changed.
+
+        // what normally happens is that the session reference should be updated to point to 'pizza',
+        // not 'loadedPizza' after the save. i.e., pizza is re-attached, by magic. And, if a single-statement request is made
+        // (e.g. by explicitly saving the Crust first), this does in fact happen.
+
+        // However, in the case of a two-statement request, the objects referenced in the second statement ('pizza') do not
+        // appear to be updated into the session cache. As a result, because the Session is now a pure-cache
+        // and no longer a read-and-merge cache, when you ask for the pizza object again, you get
+        // the cached copy of 'loadedPizza', which is now stale, and doesn't reflect the changes you just made.
+
+    }
+
+    @Test
+    public void shouldBeAbleToModifyPropertiesAndRelsWithinSingleSave() {
+
+        Pizza pizza = new Pizza();
+
+        session.save(pizza);
+
+        // detach the pizza
+        session.clear();
+
+        // NOTE: if we instead reload into our pizza object, the test will pass, but what
+        // this does is create a new object in the session. And, rather than work with
+        // the new attached object 'loadedPizza', we continue to work with the detached object, 'pizza'.
+        // this is the first condition for failure.
+        Pizza loadedPizza = session.load(Pizza.class, pizza.getId());
 
         // now we create a relationship and update a property on the detached pizza object.
         // note that we don't save the Crust first. Crust is a new object when pizza is saved, so
         // we will generate a 2-statement cypher request. This is the second condition for failure
-        pizza.setCrust(new Crust("Thick Crust"));
+        Crust crust = new Crust("Thin Crust");
+        pizza.setCrust(crust);
+        pizza.setName("Just bread");
+
+        // pizza should be dirty
+        Assert.assertTrue(((Neo4jSession) session).context().isDirty(pizza));
+        session.save(pizza);
+        // pizza should NOT be dirty - but it is, indicating it's not current in the cache.
+        Assert.assertFalse(((Neo4jSession) session).context().isDirty(pizza)); // this should pass
+
+        loadedPizza = session.load(Pizza.class, pizza.getId());
+
+        pizza.setName("Just bread");
 
-        // now, our detached pizza should be dirty, but it is not
         Assert.assertTrue(((Neo4jSession) session).context().isDirty(pizza));
     }
 
```

Comment:
Overall my thoughts about this are that I don't understand it properly. In particular I don't immediately see why the hash should be updated in some circumstances, but not in others. I think we need more tests to make this reasoning explicit. The boolean flag meaning "do this sometimes" is opaque. It would be better to analyse why we don't need to call remember at the various points it is currently invoked and avoid calling it altogether. This would make the reasoning about the invocation explicit. It would also imply moving the updating of labels to a separate method, since that always appears to be executed regardless of whether the hash is updated.